scipy/stats/_rvs_sampling.py

```diff
@@ -168,3 +172,393 @@ def rvs_ratio_uniforms(pdf, umax, vmin, vmax, size=1, c=0, random_state=None):
         i += 1
 
     return np.reshape(x, size1d)
+
+
+class FastNumericalInverse():
+    r"""
+    Object representing a Fast Numerical Inverse of a probability distribution.
+
+    The initializer of `FastNumericalInverse` accepts `dist`, a frozen instance
+    of a continuous distribution, and provides an object with methods
+    that approximate `dist.ppf` and `dist.rvs`. For most distributions,
+    these methods are faster than those of `dist` itself.
+
+    Parameters
+    ----------
+    dist : scipy.stats.rv_frozen
+        Frozen distribution for which a fast numerical inverse is desired.
+    tol : float, optional
```

Comment:
The default in UNU.RAN is 1e-10:

> As the resolution of most uniform random number sources is 2^(-32) = 2.3e-10, a value of 1.e-10 leads to an inversion algorithm that could be called exact. For most simulations slightly bigger values for the maximal error are enough as well.

Does anyone know the resolution of the numpy uniform RNG?
Does anyone know the

Reply:
@rkern Can you tell us the resolution of the numpy RNG? 

I assumed it was machine precision, which is why I suggested that we increase QMCEngine resolution from 30-bit to e.g. 52. (Should we?)

Reply:
We draw random uniform integers in [`[0, 2**53)` and divide by `2**53`](https://github.com/numpy/numpy/blob/main/numpy/random/_common.pxd#L67-L68), so ~1e-16. I don't know how much imposing that tolerance requirement is going to punish this algorithm.

Reply:
(So @rkern what do you think about increasing the QMCEngine resolution? Do you think that would be an important thing to do before the 1.7 branches, or would changing it after QMCEngine has released be OK? If it needs to be changed, I can open a separate issue.)

![image](https://user-images.githubusercontent.com/6570539/118559505-dd052000-b71c-11eb-878c-486b12fdb445.png)

Reply:
There is no particular "resolution" for `QMCEngine`s. The `2**30` that you are referring to is just a feature of the `Sobol` algorithm, none of the others, and it's not particularly a "resolution" that's worth increasing to "match" our PRNG generators. In the `Sobol` case, you can sort of squint and see the similarity between the `2**30` in it and the `2**53` in our PRNG floating-point conversion, but there's nothing in `Halton`, for example, that looks remotely like it.

You should treat the output of `QMCEngine` just like the output of our PRNGs for this calculation.

Reply:
OK. Thanks.

Reply:
Thanks for the clarification, 1e-12 indeed looks like a good default